rustls/src/client/hs.rs

```diff
@@ -683,71 +681,78 @@ impl State for ExpectServerHello {
+                        }))
+                    } else {
+                        Ok(Box::new(tls12::ExpectCCS {
+                            secrets,
+                            handshake: self.handshake,
+                            session_id: session_id,
+                            dns_name: self.dns_name,
+                            using_ems: self.using_ems,
+                            transcript: self.transcript,
+                            ticket: ReceivedTicketDetails::new(),
+                            resuming: true,
+                            cert_verified,
+                            sig_verified,
+                        }))
+                    };
                 }
-
-                let secrets =
-                    SessionSecrets::new_resume(&self.randoms, suite, &resuming.master_secret.0);
-                sess.config.key_log.log(
-                    "CLIENT_RANDOM",
-                    &secrets.randoms.client,
-                    &secrets.master_secret,
-                );
-                sess.common
-                    .start_encryption_tls12(&secrets);
-
-                // Since we're resuming, we verified the certificate and
-                // proof of possession in the prior session.
-                sess.server_cert_chain = resuming.server_cert_chain.clone();
-                let cert_verified = verify::ServerCertVerified::assertion();
-                let sig_verified = verify::HandshakeSignatureValid::assertion();
-
-                return if must_issue_new_ticket {
-                    Ok(Box::new(tls12::ExpectNewTicket {
-                        secrets,
-                        handshake: self.handshake,
-                        session_id: self.session_id,
-                        dns_name: self.dns_name,
-                        using_ems: self.using_ems,
-                        transcript: self.transcript,
-                        resuming: true,
-                        cert_verified,
-                        sig_verified,
-                    }))
-                } else {
-                    Ok(Box::new(tls12::ExpectCCS {
-                        secrets,
-                        handshake: self.handshake,
-                        session_id: self.session_id,
-                        dns_name: self.dns_name,
-                        using_ems: self.using_ems,
-                        transcript: self.transcript,
-                        ticket: ReceivedTicketDetails::new(),
-                        resuming: true,
-                        cert_verified,
-                        sig_verified,
-                    }))
-                };
+            }
+            _ => {
+                debug!("Did not resume session.");
             }
         }
 
         Ok(Box::new(tls12::ExpectCertificate {
             handshake: self.handshake,
-            session_id: self.session_id,
+            session_id: self
+                .session_id
+                .ok_or(TlsError::HandshakeNotComplete)?,
```

Comment:
I spent some time figuring out why this is here, but I can't quite figure it out. Why do we need this?

Reply:
I thought that must be true as a result of line 634, but I can make it an Option as well. It also relates to the question I asked above:

> I wasn't sure what to do when reading a zero-length SessionID off the wire. I left it as-is, but perhaps it should result in None.

I don't think this will hold if the server's session_id can be None.

Reply:
I do think reading a zero-length `SessionID` off the wire should result in `None` (IMO only non-empty `SessionID`s should be `Some`).

Reply:
OK, I will change adjust this to interpret zero-length `SessionID` messages as None.